src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -739,73 +759,88 @@ DatabaseShardImp::import(Database& source)
+            std::shared_ptr<Ledger> ledger;
+            std::uint32_t ledgerSeq;
+            std::tie(ledger, ledgerSeq, std::ignore) = loadLedgerHelper(
+                "WHERE LedgerSeq >= " + std::to_string(earliestLedgerSeq()) +
+                    " order by LedgerSeq " + sortOrder + " limit 1",
+                app_,
+                false);
+            if (!ledger || ledgerSeq == 0)
             {
                 JLOG(j_.error()) << "no suitable ledgers were found in"
                                     " the SQLite database to import";
-                return;
+                return boost::none;
             }
-        }
+            return ledgerSeq;
+        };
 
-        auto numHistShards = this->numHistoricalShards(lock);
+        // Find earliest ledger sequence stored
+        auto const earliestLedgerSeq{loadLedger()};
+        if (!earliestLedgerSeq)
+            return;
+        earliestIndex = seqToShardIndex(*earliestLedgerSeq);
 
-        // Import the shards
-        for (std::uint32_t shardIndex = earliestIndex;
-             shardIndex <= latestIndex;
-             ++shardIndex)
-        {
-            auto const pathDesignation =
-                prepareForNewShard(shardIndex, numHistShards, lock);
+        // Consider only complete shards
+        if (earliestLedgerSeq != firstLedgerSeq(earliestIndex))
+            ++earliestIndex;
 
-            if (!pathDesignation)
-                break;
+        // Find last ledger sequence stored
+        auto const latestLedgerSeq = loadLedger("desc");
+        if (!latestLedgerSeq)
+            return;
+        latestIndex = seqToShardIndex(*latestLedgerSeq);
+
+        // Consider only complete shards
+        if (latestLedgerSeq != lastLedgerSeq(latestIndex))
+            --latestIndex;
+    }
 
-            auto const needsHistoricalPath =
-                *pathDesignation == PathDesignation::historical;
+    if (latestIndex < earliestIndex)
+    {
+        JLOG(j_.error()) << "no suitable ledgers were found in"
+                            " the SQLite database to import";
+        return;
+    }
+
+    auto numHistShards = numHistoricalShards(std::lock_guard(mutex_));
+
+    JLOG(j_.debug()) << "Importing ledgers for shards " << earliestIndex
+                     << " through " << latestIndex;
+
+    {
+        std::lock_guard lock(mutex_);
+
+        importStatus_ = std::make_unique<DatabaseImportStatus>(
+            earliestIndex, latestIndex, 0);
+    }
+
+    // Import the shards
+    for (std::uint32_t shardIndex = earliestIndex; shardIndex <= latestIndex;
+         ++shardIndex)
+    {
+        if (isStopping())
+            return;
+
+        auto const pathDesignation = prepareForNewShard(
+            shardIndex, numHistShards, std::lock_guard(mutex_));
+
+        if (!pathDesignation)
+            break;
+
+        {
+            std::lock_guard lock(mutex_);
 
             // Skip if being acquired
             if (shardIndex == acquireIndex_)
```

Comment:
`prepareIndexes_` is used for shards being imported after being downloaded by the archive handler. I added a check that ensures the shards we add to that container never contain the shard currently being imported from the nodestore.

Reply:
I think I was not very clear about my concerns here.  I'll try to be more concise.  Here's the first of two concerns.

- There used to be just one operation referred to as an `import` in the `DatabaseShardImp`.  That was an import after a download from the archive handler.
- Now you're adding an operation where you `import` from the node store.

To my mind calling both of these operations just plain `import` is a recipe for confusion.  It would be great if the comments and method names distinguished the _kind_ of `import`.  So, instead of just `import`, maybe...

- `archiveImport` and
- `nodestoreImport`?

These names are just off the cuff.  You can probably do better.

--------------------------------------------------------

It looks like the second concern is not an issue.  The `importStatus_` member is being updated by `doImport()` and being checked by `DatabaseShardImp::prepareShards()`.

Reply:
Done (comments and function naming)